CriteoPublisherSdk/Sources/Criteo.m

```diff
@@ -54,17 +54,31 @@ + (void)initialize {
   }
 }
 
+static Criteo *sharedInstance = nil;
+static dispatch_once_t onceToken;
+
 + (instancetype)sharedCriteo {
-  static Criteo *sharedInstance = nil;
-  static dispatch_once_t onceToken;
+  // The initialization step of the SDK involves logger which need the sharedCriteo.
+  // To handle the circular dependency, the pure instantiation is separated from the initialization.
+  __block BOOL wasInstantiated = NO;
+
   dispatch_once(&onceToken, ^{
-    sharedInstance = [self criteo];
-    CRLogInfo(@"Initialization", @"Singleton was initialized");
+    sharedInstance = [[self alloc] initWithDependencyProvider:CR_DependencyProvider.new];
```

Comment:
Done.

I have a question about this: is it really possible to get an exception while allocating an object ? ObjectiveC seems to come from the C, which can clearly fail to `malloc` (for instance because of an undefined behavior when moon is in front of Saturn and the temperature of CPU is 42°C). Although, ObjectiveC is not as raw as C on memory management. So I'm wondering if failure of alloc in Obj-C isn't it more like the `OutOfMemoryError` from Java where, if such thing happens, generally you should not catch them because you're in a big trouble anyway.
Well I would like to get your point of view on memory error handling on Obj-C :)

Reply:
Good question. Exceptions in Objective C are quite rare and mostly for logic cases like "array out of bound". Never saw an alloc failure which is normal as it is mostly size calculation then a malloc. Malloc is not designed to fail, when memory goes wrong, like out of memory, the OS takes over. iOS for instance has a "watchdog" process that will [signal  processes to ask them to release memory](https://developer.apple.com/documentation/xcode/improving_your_app_s_performance/reducing_your_app_s_memory_use/responding_to_low-memory_warnings), then will kill processes.